File: kotlin/src/main/kotlin/io/github/konfigur8/Property.kt
package io.github.konfigur8

import java.util.*

class Property<T>(val name: String, val deserialize: (String) -> T, val serialize: (T) -> String = { it: T -> it.toString() }, val exposeMode: ExposeMode = ExposeMode.Public) {
    override fun toString() = name

    companion object {
        fun string(name: String, exposeMode: ExposeMode = ExposeMode.Public) = Property(name, { it }, exposeMode = exposeMode)

        fun int(name: String, exposeMode: ExposeMode = ExposeMode.Public) = Property(name, { it.toInt() }, exposeMode = exposeMode)

        fun long(name: String, exposeMode: ExposeMode = ExposeMode.Public) = Property(name, { it.toLong() }, exposeMode = exposeMode)

        fun bool(name: String, exposeMode: ExposeMode = ExposeMode.Public) = Property(name, { it.toBoolean() }, exposeMode = exposeMode)

        fun character(name: String, exposeMode: ExposeMode = ExposeMode.Public) = Property(name, { it.first() }, exposeMode = exposeMode)

        fun uuid(name: String, exposeMode: ExposeMode = ExposeMode.Public) = Property(name, { UUID.fromString(it) }, exposeMode = exposeMode)
    }
}

File: javascript/src/Property.ts
import {ExposeMode} from "./ExposeMode";

export class Property<T> {
    public readonly name: string;
    public readonly deserialize: (value: string) => T;
    public readonly serialize: (value: T) => string;
    public readonly exposeMode: ExposeMode;

    constructor(
        name: string,
        deserialize: (value: string) => T,
        serialize: (value: T) => string = (value: T) => value.toString(),
        exposeMode: ExposeMode = ExposeMode.Public
    ) {
        this.name = name;
        this.deserialize = deserialize;
        this.serialize = serialize;
        this.exposeMode = exposeMode;
    }

    static string(name: string, exposeMode: ExposeMode = ExposeMode.Public) {
        return new Property(
            name,
            (value) => value,
            (value) => value.toString(),
            exposeMode
        )
    }

    static int(name: string, exposeMode: ExposeMode = ExposeMode.Public) {
        return new Property(
            name,
            (value) => parseInt(value),
            (value) => value.toString(),
            exposeMode
        )
    }

    static float(name: string, exposeMode: ExposeMode = ExposeMode.Public) {
        return new Property(
            name,
            (value) => parseFloat(value),
            (value) => value.toString(),
            exposeMode
        )
    }

    static character(name: string, exposeMode: ExposeMode = ExposeMode.Public) {
        return new Property(
            name,
            (value) => value.charAt(0),
            (value) => value.toString(),
            exposeMode
        )
    }
}
